File: ts/test/lookup.test.ts
/**
 * Lookup conformance: every case in conformance/golden/lookup.json must
 * resolve to the same kind, or fail with the same (at, segment, reason)
 * triple, as the Go reference. Plus port-local cases the golden cannot
 * host (dangling ref, name rules).
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";
import { create, fromJson } from "@bufbuild/protobuf";
import { describe, expect, it } from "vitest";
import { checkDescriptor } from "../src/descriptor.js";
import { SchemaSchema } from "../src/gen/schemapb/schema_pb.js";
import { LookupError, listItems, lookupPath } from "../src/lookup.js";
import { kindName } from "../src/render.js";

const goldenDir = join(__dirname, "..", "..", "conformance", "golden");

interface LookupCase {
  path: string;
  kind?: string;
  items?: string[];
  error?: { at: string; segment: string; reason: string };
}

describe("lookup conformance", () => {
  const schema = fromJson(
    SchemaSchema,
    JSON.parse(readFileSync(join(goldenDir, "full-schema.json"), "utf8")),
  );
  const cases: LookupCase[] = JSON.parse(
    readFileSync(join(goldenDir, "lookup.json"), "utf8"),
  ).cases;

  it.each(cases)("$path", (c) => {
    if (c.error !== undefined) {
      try {
        lookupPath(schema, c.path);
        expect.fail(`lookupPath(${c.path}) resolved, want ${c.error.reason}`);
      } catch (err) {
        expect(err).toBeInstanceOf(LookupError);
        const le = err as LookupError;
        expect({ at: le.at, segment: le.segment, reason: le.reason }).toEqual(c.error);
      }
      return;
    }
    const f = lookupPath(schema, c.path);
    expect(kindName(f)).toBe(c.kind);
    expect(listItems(f).map(kindName)).toEqual(c.items ?? []);
  });
});

describe("field name rules", () => {
  const schemaWithName = (name: string) =>
    create(SchemaSchema, {
      id: { namespace: "t", name: "names" },
      fields: [{ name, kind: { case: "string", value: {} } }],
    });

  it.each(["a.b", "my-field", "1st", "in", "true", "while"])("rejects %s", (bad) => {
    expect(checkDescriptor(schemaWithName(bad)).length).toBeGreaterThan(0);
  });

  it.each(["snake_case", "camelCase", "_x", "a1"])("accepts %s", (good) => {
    expect(checkDescriptor(schemaWithName(good))).toEqual([]);
  });
});

File: go/schemapb/lookup_test.go
package schemapb_test

import (
	"errors"
	"testing"

	"github.com/gopherex/schemapb/go/schemapb"
)

// Cases the kitchen-sink golden cannot host: a broken ref (the golden
// schema must stay valid), typed-segment Lookup, error texts, name rules.

func lookupFixture(t *testing.T) *schemapb.Schema {
	t.Helper()

	s, err := schemapb.NewSchema(schemapb.ID("t", "lookup", schemapb.Ver(1, 0, 0))).
		Fields(
			schemapb.Str("name"),
			schemapb.Object("cfg", schemapb.Bool("on")),
		).
		Build()
	if err != nil {
		t.Fatal(err)
	}

	return s
}

func TestLookupTypedSegments(t *testing.T) {
	t.Parallel()

	s := lookupFixture(t)

	f, err := s.Lookup(schemapb.FieldName("cfg"), schemapb.FieldName("on"))
	if err != nil {
		t.Fatal(err)
	}

	if got := schemapb.KindName(f); got != schemapb.KindBool {
		t.Fatalf("kind = %q, want bool", got)
	}
}

func TestLookupErrorText(t *testing.T) {
	t.Parallel()

	s := lookupFixture(t)

	_, err := s.LookupPath("cfg.off")

	var lerr *schemapb.LookupError
	if !errors.As(err, &lerr) {
		t.Fatalf("want LookupError, got %v", err)
	}

	if lerr.At != "cfg" || lerr.Segment != "off" || lerr.Reason != schemapb.LookupNotFound {
		t.Fatalf("wrong error fields: %+v", lerr)
	}

	const want = `schemapb: lookup: no field "off" in "cfg"`
	if lerr.Error() != want {
		t.Fatalf("error text = %q, want %q", lerr.Error(), want)
	}
}

func TestLookupUnknownRef(t *testing.T) {
	t.Parallel()

	// Built by hand: builders/Compile would refuse the dangling ref.
	s := &schemapb.Schema{
		Fields: []*schemapb.Schema_Field{{
			Name: "vol",
			Kind: &schemapb.Schema_Field_Ref_{Ref: &schemapb.Schema_Field_Ref{
				Target: &schemapb.Schema_Field_Ref_Name{Name: "missing"},
			}},
		}},
	}

	_, err := s.LookupPath("vol.path")

	var lerr *schemapb.LookupError
	if !errors.As(err, &lerr) || lerr.Reason != schemapb.LookupUnknownRef {
		t.Fatalf("want unknown_ref, got %v", err)
	}
}

func TestLookupEmptyPath(t *testing.T) {
	t.Parallel()

	s := lookupFixture(t)

	for _, try := range []func() (*schemapb.Schema_Field, error){
		func() (*schemapb.Schema_Field, error) { return s.Lookup() },
		func() (*schemapb.Schema_Field, error) { return s.LookupPath("") },
	} {
		_, err := try()

		var lerr *schemapb.LookupError
		if !errors.As(err, &lerr) || lerr.Reason != schemapb.LookupEmptyPath {
			t.Fatalf("want empty_path, got %v", err)
		}
	}
}

func TestFieldNameRules(t *testing.T) {
	t.Parallel()

	for _, bad := range []string{"a.b", "my-field", "1st", "in", "true", "while"} {
		s := &schemapb.Schema{
			Id: schemapb.ID("t", "names", schemapb.Ver(1, 0, 0)),
			Fields: []*schemapb.Schema_Field{{
				Name: bad,
				Kind: &schemapb.Schema_Field_String_{String_: &schemapb.Schema_Field_String{}},
			}},
		}

		if _, err := schemapb.Compile(s); err == nil {
			t.Errorf("field name %q compiled, want descriptor error", bad)
		}
	}

	// snake_case and camelCase both stay legal.
	for _, good := range []string{"snake_case", "camelCase", "_x", "a1"} {
		s := &schemapb.Schema{
			Id: schemapb.ID("t", "names", schemapb.Ver(1, 0, 0)),
			Fields: []*schemapb.Schema_Field{{
				Name: good,
				Kind: &schemapb.Schema_Field_String_{String_: &schemapb.Schema_Field_String{}},
			}},
		}

		if _, err := schemapb.Compile(s); err != nil {
			t.Errorf("field name %q rejected: %v", good, err)
		}
	}
}
